src/emulator/hub.ts

```diff
@@ -0,0 +1,158 @@
+import * as http from "http";
+import * as express from "express";
+import * as os from "os";
+import * as fs from "fs";
+import * as path from "path";
+import * as bodyParser from "body-parser";
+
+import * as utils from "../utils";
+import * as logger from "../logger";
+import { Constants } from "./constants";
+import { Emulators, EmulatorInstance, EmulatorInfo, IMPORT_EXPORT_EMULATORS } from "./types";
+import { HubExport } from "./hubExport";
+
+// We use the CLI version from package.json
+const pkg = require("../../package.json");
+const version = pkg.version;
+
+export interface Locator {
+  version: string;
+  host: string;
+  port: number;
+}
+
+export interface EmulatorHubArgs {
+  projectId: string;
+  port?: number;
+  host?: string;
+}
+
+export class EmulatorHub implements EmulatorInstance {
+  static PATH_EXPORT = "/_admin/export";
+
+  /**
+   * Given a project ID, find and read the Locator file for the emulator hub.
+   * This is useful so that multiple copies of the Firebase CLI can discover
+   * each other.
+   */
+  static readLocatorFile(projectId: string): Locator | undefined {
+    const locatorPath = this.getLocatorFilePath(projectId);
+    if (!fs.existsSync(locatorPath)) {
+      return undefined;
+    }
+
+    const data = fs.readFileSync(locatorPath).toString();
+    const locator = JSON.parse(data) as Locator;
+
+    if (locator.version !== version) {
+      logger.debug(`Found locator with mismatched version, ignoring: ${JSON.stringify(locator)}`);
+      return undefined;
+    }
+
+    return locator;
+  }
+
+  static getLocatorFilePath(projectId: string): string {
+    const dir = os.tmpdir();
+    const filename = `hub-${projectId}.json`;
```

Comment:
So... say that I have two git clones of the same project and I modified the ports in one of them, so that I can emulate two versions of my app and compare them side to side. They surely have the same project ID. What will happen? Do we even support this? If not, how do we better surface the error and tell people like me what to do?

Reply:
Right now the hub started second would clobber the first one.  I'll add a warning but not sure what else I can do.

Reply:
I think this should be enough for now.